dist/images/README_IMAGE.md

```diff
@@ -0,0 +1,252 @@
+containers:
+- name: run-ovn-northd
+  image: {{.OvnImage}}
+  command: ["/root/ovnkube.sh", "run-ovn-northd"]
+```
+
+The interface includes the command names (e.g., "run-ovn-northd"),
+a set of environment variables, configMaps, and a set of files that mounted
+into the container. Over time the interface evolves and the daemonsets need
+to be updated as well. To assist in this the OVN_DAEMONSET_VERSION environment
+variable is used by the daemonset to specify the version of the interface it
+is using. The ovnkube.sh script notes the version and operates accordingly.
+
+
+## Image Component Parts
+
+Images created from this repo are for development, debug and local testing. In a product
+context the product's build environment will build and push the image to a
+suitable image repository.
+
+The image includes:
+* The ovnkube.sh script
+* The ovnkube binary
+* openvswitch (ovs) and open virtual network (ovn)
+* a base for the image and dependency packages.
+
+Some notes on the components:
+* The ovnkube.sh script provides the interface between the daemonsets and the image.
+  Because the interface evolves over time it is versioned. The script supports two
+  consecutive versions of the interface and older versions are depricated. This
+  permits evolving the interface while not breaking existing users.
+
+* The ovnkube binary is built in ../../go_controller and copied to this directory.
+
+* ovs/ovn evolves over time and a suitable version needs to be included in the image.
+  As of this writing ovs/ovn 2.11, or higher, is required. It is up to the image
+  builder to include a suitable version.
+
+* The base image is from any of many distributions. It is up to the image biulder to
+  select a suitable base. Depending on the base image, additional packages may be
+  needed to satisfy dependencies. There are sample Dockerfiles included here
+  that can be used as examples and to build development images. And a Makefile
+  to assist in the image build.
+
+There is some complexity in building the image. Each base may require additional
+packages to satisfy dependencies. In particular, a suitable version of ovs/ovn
+needs to be included. Also, automated build environments may impose
+additional requirements. The Dockerfiles in this repo will build images suitable for
+development and debug.
+
+The image must be stored in an image repository that can be accessed by pods in the
+cluster. The products that use ovn-kubernetes usually build the image and push it to
+a selected repository. The "image:" name in the daemonsets points to where the image
+was pushed.
+
+# Interface Versions
+
+Every container needs to set an environment variable that specifies the
+interface version that it is using.
+
+```
+env:
+- name: OVN_DAEMONSET_VERSION
+  value: "4"
+```
+At present version "4" and "3" are supported.
+
+# Version 4 interface
+
+## ovn-config configMap
+
+The ovn-config configMap is mounted into the container in directory
+```
+/var/run/ovn-config
```

Comment:
For the `ovn-config` ConfigMap, I don't think there's any advantage to doing it this way rather than sticking with environment variables as in v3. Especially since `ovnkube.sh` still needs to have the code to deal with the environment variables anyway...

Reply:
@danwinship I think the one big reason we wanted a ConfigMap was to reduce the size/churn of ovnkube.sh. With a CM we can set many options, including new ones, from the DaemonSet without having to rev ovnkube.sh for every small change. And honestly I'd rather kill the environment variables so that we only have 2 ways of delivering config (CLI and config file) rather than 3 (CLI, file, env). ConfigMap also allows us to dynamically change config in the future (not that we'd use it now or soon).

Reply:
No, I mean, there are two ConfigMaps here; `ovnkube-config` gets passed as a file to `/usr/bin/ovnkube` and that part totally makes sense. But `ovn-config` is the stuff that `ovnkube.sh` itself uses directly, and it just copies all of its contents into environment variables. Except that we already had code to deal with that in v3 by having the pod spec copy the configmap into environment variables, and it's not clear what the v4 code gains us there.

Now if you wanted to have `ovnkube.sh` parse the `cluster-subnets`, `service-cidr`, `ovn-config-namespace`, and `apiserver` out of `ovnkube-config`, and then get rid of the `ovn-config` ConfigMap entirely, that would make sense.